### DISCUSSION TITLE
Domain Model

### DISCUSSION THREAD

**Discussion by pablogugarcia** [SuperNerdIT/telepathic#14] [Ideas]
Domain model

This model is for comprehension of the sistem not necessary a class diagram o software representation

``` mermaid 
classDiagram
class Topic { }
class Message { }
class UserSubscriber { }
class UserSender { }

UserSender "*" -- "*" Message : Emits
UserSubscriber "*" -- "*" Topic : Subscribe-To
Message "1" -- "1" Topic : Contains

```

Use cases:

Easy
- User subscribes to a topic and receives all messages sent to a that topic since the subscription
- User sends a message with a topic 

Medium
- Broker internally have retries for failed deliveries

Hard
- User can sent a message to a topic and ensure that the message arribes to all the subscribed user ( E.g transactional = true flag, after all user receives the message the broker makes a "commit" )
-  User can know the percentages of users that actually received a message with a certain id.

How can we improve it ? @LautaroJayat

**Comment by LautaroJayat** [comment]
The only part that I'm not pretty sure about is the link where `Topic` listens to an `Event`.
I would rather think that the topic is part of the message.
Maybe we can about it in the following way:
1. The user suscribes to a topic by interacting with some module
2. The module registers the petition and starts listening to the events
3. The event contains a message with some topic.

But let keep thinking about it.
